sdk/keyvault/azure-security-keyvault-administration/src/test/java/com/azure/security/keyvault/administration/KeyVaultCredentialPolicyTest.java

```diff
@@ -43,85 +45,120 @@ public class KeyVaultCredentialPolicyTest {
+        MockHttpResponse unauthorizedResponseWithHeaderAndClaims = new MockHttpResponse(
+            new HttpRequest(HttpMethod.GET, "https://azure.com"), 401,
+            new HttpHeaders().set(HttpHeaderName.WWW_AUTHENTICATE, AUTHENTICATE_HEADER_WITH_CLAIMS));
 
+        this.unauthorizedHttpResponseWithWrongStatusCode = unauthorizedResponseWithWrongStatusCode;
         this.unauthorizedHttpResponseWithHeader = unauthorizedResponseWithHeader;
         this.unauthorizedHttpResponseWithoutHeader = unauthorizedResponseWithoutHeader;
-        this.callContext = plainContext;
-        this.differentScopeContext = differentScopeContext;
+        this.unauthorizedHttpResponseWithHeaderAndClaims = unauthorizedResponseWithHeaderAndClaims;
+        this.callContext = createCallContext(request, Context.NONE);
+        this.differentScopeContext = createCallContext(requestWithDifferentScope, Context.NONE);
         this.credential = new BasicAuthenticationCredential("user", "fakePasswordPlaceholder");
-        this.testContext = testContext;
-        this.bodyContext = bodyContext;
-        this.bodyFluxContext = bodyFluxContext;
+        this.testContext = createCallContext(request, Context.NONE);
+        this.bodyContext = createCallContext(request, bodyContextContext);
+        this.bodyFluxContext = createCallContext(request, bodyFluxContextContext);
+    }
 
+    @AfterEach
+    public void cleanup() {
+        KeyVaultCredentialPolicy.clearCache();
     }
 
-    private static HttpPipelineCallContext createMockContext(HttpRequest request,
-        Consumer<HttpPipelineCallContext> contextSetter) {
-        AtomicReference<HttpPipelineCallContext> capturedContext = new AtomicReference<>();
-        HttpPipeline httpPipeline = new HttpPipelineBuilder()
-            .policies((httpPipelineCallContext, httpPipelineNextPolicy) -> {
-                if (contextSetter != null) {
-                    contextSetter.accept(httpPipelineCallContext);
-                }
-
-                capturedContext.set(httpPipelineCallContext);
-                return Mono.empty();
-            })
-            .httpClient(ignored -> Mono.empty())
+    @SyncAsyncTest
+    public void onNon401ErrorResponse() {
+        KeyVaultCredentialPolicy policy = new KeyVaultCredentialPolicy(this.credential, false);
+
+        HttpPipeline pipeline = new HttpPipelineBuilder()
+            .policies(policy)
+            .httpClient(ignored -> Mono.just(unauthorizedHttpResponseWithWrongStatusCode))
             .build();
 
-        httpPipeline.send(request).block();
+        SyncAsyncExtension.execute(
+            () -> pipeline.sendSync(this.callContext.getHttpRequest(), this.callContext.getContext()),
+            () -> pipeline.send(this.callContext.getHttpRequest(), this.callContext.getContext())
+        );
+
+        assertNull(this.callContext.getHttpRequest().getHeaders().get(AUTHORIZATION));
```

Comment:
Is this asserting the policy doesn't authorize its first request or that it doesn't retry a 500 with authorization?

Reply:
Well, right now it's testing it does not handle a 500 response even if it happened to contain a challenge. I guess I could write a separate test that first gets a 401 with a challenge and then a 500 with a challenge and claims, but I doubt we should be exhaustively testing if an `if (response.getStatus() == 401)` condition works.

Reply:
👍 I'm comfortable not testing that because a 500 response with an auth challenge would be a bizarre server bug. It doesn't make sense semantically and there's no scenario for it, so we don't need to defend against it

Reply:
It does not care about if it's a first request or a subsequent one, only about what happens when we get a response with a status code that's not `401`.